test/test_sox_compatibility.py

```diff
@@ -9,9 +9,27 @@
 
 
 @common_utils.skipIfNoSoxBackend
-class TestFunctionalFiltering(common_utils.TorchaudioTestCase):
+class TestFunctionalFiltering(common_utils.TempDirMixin, common_utils.TorchaudioTestCase):
     backend = 'sox'
 
+    def setUp(self):
+        common_utils.TempDirMixin.setUp(self)
+        common_utils.TorchaudioTestCase.setUp(self)
```

Comment:
So with the current order, wouldn't `TorchaudioTestCase.setUp()` never get called? Which means that `TestBaseMixin.setUp()` would never get called? which means `set_audio_backend(self.backend)` never gets called?

What if in the future we add more functionality that we want to rely on with other mixins?  Wouldn't we want to have this robust to that?  Thoughts?

Reply:
with `class TestFo(TempDirMixin, TorchaudioTestCase)`, the following happens.

1. Test framework tries to invoke `setUp` on `TestFoo`.

Case 1.  If `TestFoo` has `setUp` method defined
2.1. TestFoo.setUp is called.
2.2. If `TestFoo.setUp` has `super().setUp()`, then `setUp` on super class is called.
2.3. It keeps resolving super class methods as long as `super().setUp()` is defined.  (supposed to stop at `unittest.TestCase`).

Case 2. If `TestFoo` does not have `setUp` method defined
2. According to MRO rule, it invokes `TempDirMixin.setUp`.
3 Inside of `TempDirMixin.setUp`, `super().setUp()` is executed.
4. `super()` finds `TorchaudioTestCase` as super class so `TorchaudioTestCase.setUp` is invoked.
5.. 3. & 4. is repeated as long as `super().setUp()` is defined in `setUp` method of any class (supposed to stop at `unittest.TestCase`).


Suppose all the `setUp` methods have  call to `super().setUp()`

For case 1 the following methods are called.
`TestFoo.setUp()` -> `TempDirMixin.setUp()` -> `TorchaudioTestCase.setUp()` -> `torch.testing._internal.common_utils.TestCase.setUp()` -> some other classes -> `unites.TestCase`.

For case 2. the following methods are called.
`TempDirMixin.setUp()` -> `TorchaudioTestCase.setUp()` -> `torch.testing._internal.common_utils.TestCase.setUp()` -> some other classes -> `unites.TestCase`.

`MRO` defines where your method starts, but the chain of superclass calls depend on the implementation. (`super().setUp()`).

And the design of MixIn here is that they add features to base `TorchaudioTestCase`, so it makes sense to start the call chain from `Mixin`. (Because `TorchaudioTestCase` class is not aware of `Mixin`.)

When you have multiple MixIn, and you implement the test class as follow;

`class MyTest(Mixin2, Mixin1, TorchaudioTestCase)`

Then the inheritance looks like this;

```
TorchaudioTestCase
↑ super()
Mixin1
↑super()
Mixin2
```

and the MRO starts from `Mixin2`.

Reply:
It is confusing but when you do multiple inheritance in Python, you need to put most parent class on the right side. I often get confused on this one and had to google every time. That will give you a right inheritance order, and with that, `super()` should work as expected.